internal/blockprocessor/committer.go

```diff
@@ -369,21 +382,44 @@ func constructDBEntriesForDBAdminTx(tx *types.DBAdministrationTx, version *types
 			value = v
 		}
 
-		db := &worldstate.KVWithMetadata{
+		createDB := &worldstate.KVWithMetadata{
 			Key:   dbName,
 			Value: value,
 			Metadata: &types.Metadata{
 				Version: version,
 			},
 		}
-		indexForExistingDBs = append(indexForExistingDBs, db)
+		indexForExistingDBs = append(indexForExistingDBs, createDB)
```

Comment:
Questions:
1. If the database already contains an index, that would overwrite its definition and render the old one stale; maybe it makes more sense to delete it and start fresh? or if the update just adds an index, it makes sense to add keys to it;
2. For a database that already contains data, a new index starts as stale.

How do you plan to handle the index building process, in terms of determining when an index db is in synch with the database it refers to?

One idea is to have the "value" of the \_index\_ database (currently empty I see) to reflect its state - stale, in-synch, etc. Just an idea, you might have thought of a better one... but I would like to learn what the plan is...

Reply:
For this PR, we are handling only index definition creation/updation events. Further, as stated in this issue https://github.com/IBM-Blockchain/bcdb-server/issues/14, we are considering the database to be empty. To address your Q1, we have https://github.com/IBM-Blockchain/bcdb-server/issues/15 and for Q2, we have https://github.com/IBM-Blockchain/bcdb-server/issues/21

When the database is not empty and a new index is created or an existing index is updated, we need to build index entries in the background. There are multiple scenarios and kind of complex. For example, consider the following scenarios:

1. **Database is not empty and new index is being defined** -- we need to perform an open-ended range query on the whole database and build index entry in batch, write the index entries to the index database and have a pointer to record till what key we have processed the index. Further, there can be new block commits and new values might get added. We need to keep track of them (even during node failure) and consider them once the existing data are processed. Note that the range query in leveldb acquires a snapshot and new writes are not visible. I am ignoring many corner cases here to keep it brief. In short, we need to store (1) a flag to denote whether the DB is in sync, (2) if being synched, what is the last key processed for the existing data; (3) if being synced, new keys committed while synching the index entries (just a list of block number is adequate); (4) if synched, what is the last processed block.
2. **Database is not empty and the existing index is being updated while the existing index entries are in sync with the database**
3. **Database is not empty and the existing index is being updated while the existing index entries are not in sync with the database**

Scenarios 2 and 3 are complex. Hence, we need to maintain more states to see (1) whether the index entries are in sync; (2) whether node recovery after failure can restart where it left off, etc...

If you prefer to merge this PR after having this feature end-to-end, I am okay with it.

Reply:
>If you prefer to merge this PR after having this feature end-to-end, I am okay with it.

No, my goal is to understand the underlying assumptions of the first phase of implementation. Clearly the scenarios described above are complex and require a lot of work. hence I assume that you are shooting for a simpler milestone to begin with. I am ok with that, just want to learn what it is though.